### ISSUE TITLE
How to handle "simple" retina images

### ISSUE THREAD

**Issue by idflood** [mindkomm/timmy#14]
On some website it is sometimes useful to have an image preset with just a retina version. For example a profile thumb with default of 100x100 and a retina version of 200x200.

Ideally we should be able to return something like:

    <img srcset="normal.jpg 1x, retina.jpg 2x" .. />

It seems that currently Timmy only output the image 'w'. So I tried to create 2 image presets and use that to render a picture element. The presets would look like this:

    'card' => [
        'resize' => [100, 100],
      ],
      'card@2x' => [
        'resize' => [200, 200],
      ],

And then render them with something like this:

    return '<picture>
      <source ' . get_timber_image_responsive_src($image, $preset . '@2x') . ' media="(min-device-pixel-ratio: 1.5), (min-resolution: 144dpi)">
      <source ' . get_timber_image_responsive_src($image, $preset) . '>
      <img srcset="' . get_timber_image_src($image, $preset) . '" alt="' . $image['alt'] . '"' . $attributes_str .' />
    </picture>';

This gives us:

    <picture>
      <source src="unsplash-200x200-c-default.jpg" media="(min-device-pixel-ratio: 1.5), (min-resolution: 144dpi)">
      <source src="unsplash-100x100-c-default.jpg">
      <img srcset="unsplash-100x100-c-default.jpg" alt="">
    </picture>

Notice that the `source` elements have an `src` element and not an `srcset`. Eventually chrome and firefox and probably other browser doesn't seem to support `src` attributes on the `source` element. Only the standard image is loaded, even on retina display.

The use of src instead of srcset if there is only one url is defined here: https://github.com/mindkomm/timmy/blob/master/functions-images.php#L319

**Comment by idflood** [comment]
Not sure yet about the best solution and if this would cover all situations but ideally the image preset would look like this:

    'card' => [
      'resize' => [100, 100],
      'srcset' => [2],
    ],

And we could use something like this, with a new method if needed:

    <img srcset="' . get_timber_image_srcset($image, $preset) . '" alt="' . $image['alt'] . '"' . $attributes_str .' />

**Comment by gchtr** [comment]
Hey @idflood 

Yeah, `picture` elements are not that well supported with Timmy yet, because they are quite hard to automate. You can go different paths with it and I thought it might be easier to just write out the markup you want instead of adding a lot of configuration to Timmy.

I guess as a workaround for now you could use the `get_timber_image_src()` function instead of `get_timber_image_responsive_src()`:

```php
return '<picture>
    <source src="' . get_timber_image_src($image, $preset . '@2x') . '" media="(min-device-pixel-ratio: 1.5), (min-resolution: 144dpi)">
    <source src="' . get_timber_image_src($image, $preset) . '">
    <img srcset="' . get_timber_image_src($image, $preset) . '" alt="' . $image['alt'] . '"' . $attributes_str .' />
</picture>';
```

Coming back to your proposal. When you have a configuration like the following:

```php
'card' => [
   'resize' => [100, 100],
    'srcset' => [2],
],
```

You would expect it to output the srcset with a `2x` descriptor, right? I think we could do that, because it’s not the same as `'srcset' => [ [ 2 ] ]`, which would double the size of image.

Or maybe we could just use `2x`, or `1.5x`, to make it more explicit what you want to do:

```php
'card' => [
    'resize' => [100, 100],
    'srcset' => [ '2x', '1.5x' ],
],
```

Adding `get_timber_image_srcset()` function is also a good idea 😎! It would then work like this:

- `get_timber_image_responsive()` – Gives you the automatic responsive markup.
- `get_timber_image_srcset()` – Gives you the value for a `srcset` attribute.
- `get_timber_image_src()` – Gives you the value for a `src` attribute.

Would you agree? Would that make sense?

**Comment by idflood** [comment]
Hi @gchtr ,

Thanks for the super quick feedback and obviously this great little library : )

I think you are right, the proposal is probably too simple to cover all use cases and doesn't provide enough informations to automatically render the desired markup.

Your proposition looks great to me, at least it will provide all the tools needed to build most picture markup 👍

**Comment by idflood** [comment]
Maybe one missing element is the ability to retrieve the single src of a size within srcset. Something like `get_timber_image_srcset_url($image, 'card', 2)`. So for my initial proposition it would look like this:

    <img srcset="' . get_timber_image_src($image, $preset) . ' 1x, ' . get_timber_image_srcset_url($image, $preset, 2) . ' 2x" alt="' . $image['alt'] . '"' . $attributes_str .' />

What do you think?

**Comment by gchtr** [comment]
Hey @idflood 

Sorry for answering so late. You know how it is… There’s always lots of other stuff that is more pressing. I finally had a similar use case in a project, so I could work on this.

I made a new pull request in #25 that adds a new `get_timber_image_srcset()` function. In addition with the newly supported notation for `srcset` in your image configuration, it would work like this for you.

Take the following image configuration:

```php
'card' => [
    'resize' => [ 100, 100 ],
    'srcset' => [ '1x', '2x' ]
],
```

And the following output:

```php
echo '<img srcset="' . get_timber_image_srcset( $image, $preset ) . '" />';
```

That would result in the following HTML:

```html
<img srcset="card-100.jpg 1x, card-200.jpg 2x" />
```

So I think you wouldn’t actually need a `get_timber_image_srcset_url()` function. Does that make sense? Would that work for you?

**Comment by gchtr** [comment]
Fixed in #25.
